### PR TITLE
Compile void# to unit instead of wrapped error

### PR DESCRIPTION
**Pull request by kk-hainq**
Proof-of-concept improvement for what David's sharp eyes pointed out in #4212.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs [hunk 1]
```diff
@@ -162,6 +162,7 @@ type Compiling uni fun m =
     , MonadWriter CoverageIndex m
     , PLC.GShow uni, PLC.GEq uni
     , PLC.ToBuiltinMeaning uni fun
+    , uni `PLC.Contains` ()
```
Comment (kk-hainq) [inline plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs:165]: Hopefully, this is not fighting a hack with another hack... Personally, I think it's fine demanding a unit for any variant.

Comment (michaelpj) [reply plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs:165]: This is partly what `CompilingDefault` is for, although that's a bit overkill in this case. For now I'd say just use `CompilingDefault` in `compileTyCon` and I can worry about chopping up `CompilingDefault` at some point.

File: plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs [hunk 2]
```diff
@@ -601,8 +601,8 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
         -- C# is just a wrapper around a literal
         GHC.Var (GHC.idDetails -> GHC.DataConWorkId dc) `GHC.App` arg | dc == GHC.charDataCon -> compileExpr arg
 
-        -- void# - values of type void get represented as error, since they should be unreachable
-        GHC.Var n | n == GHC.voidPrimId || n == GHC.voidArgId -> errorFunc
+        -- void# - values of type Void# get represented as unit
```
Comment (michaelpj) [inline plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:604]:
```suggestion
        -- void# - Surprisingly, `Void#` is actually more like `Unit` than `Void`, so we represent it as such
```

**Review by michaelpj — APPROVED** [review]
Looks plausible!

File: plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs (hunk 1 above)
Comment (kk-hainq) [reply plutus-tx-plugin/src/PlutusTx/Compiler/Types.hs]: The `CompilingDefault` change is more pervasive but looking at the surrounding code it is indeed better. I have updated as suggested.

File: plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs (hunk 2 above)
Comment (kk-hainq) [reply plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs]: Updated as suggested in both places.

**Comment by michaelpj** [comment]
Great, thanks!

**Pull request merged**